File: parallel-programming-with-ms-dotnet/CSharp/Chapter2/CreditReview/Program.cs
// ===============================================================================
//  Microsoft patterns & practices
//  Parallel Programming Guide
// ===============================================================================
//  Copyright © Microsoft Corporation.  All rights reserved.
//  This code released under the terms of the 
//  Microsoft patterns & practices license (http://parallelpatterns.codeplex.com/license).
// ===============================================================================

using System;
using System.Globalization;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Practices.ParallelGuideSamples.Utilities;

namespace Microsoft.Practices.ParallelGuideSamples.CreditReview
{
    class Program
    {
        private const int NumberOfMonths = 3;

        static void UpdatePredictionsSequential(AccountRepository accounts)
        {
            foreach (Account account in accounts.AllAccounts)
            {
                Trend trend = SampleUtilities.Fit(account.Balance);
                double prediction = trend.Predict(account.Balance.Length + NumberOfMonths); 
                account.SeqPrediction = prediction;
                account.SeqWarning = prediction < account.Overdraft;
            }
        }

        static void UpdatePredictionsParallel(AccountRepository accounts)
        {
            Parallel.ForEach(accounts.AllAccounts, account =>
            {
                Trend trend = SampleUtilities.Fit(account.Balance);
                double prediction = trend.Predict(account.Balance.Length + NumberOfMonths);
                account.ParPrediction = prediction;
                account.ParWarning = prediction < account.Overdraft;
            });
        }

        static void UpdatePredictionsPlinq(AccountRepository accounts)
        {            
            accounts.AllAccounts
                .AsParallel()
                .ForAll(account =>
                    {
                        Trend trend = SampleUtilities.Fit(account.Balance);
                        double prediction = trend.Predict(account.Balance.Length + NumberOfMonths);
                        account.PlinqPrediction = prediction;
                        account.PlinqWarning = prediction < account.Overdraft;         
                    });
        }

        /// <summary>
        /// Usage: CreditReview n, optional n is number of customers, use 100,000+ for meaningful timings
        /// </summary>
        static void Main(string[] args)
        {
            Console.WriteLine("Credit Review Sample\n");
#if DEBUG
            Console.WriteLine("For most accurate timing results, use Release build.\n");
#endif

            Random random = new Random(1);  // seed 1 makes runs reproducible

            // Defaults for data generation, may override some on command line
            int months = 36;
            int customerCount = 1000000; // for data runs make big enough for significant timing measurements
            Trend goodBalance = new Trend { Slope = 0.0, Intercept = 0.0 };
            Trend badBalance = new Trend { Slope = -150.0, Intercept = 0.0 };
            const double variation = 100.0;
            const double overdraft = -1000.0; // Default overdraft limit

            // Printed table of results
            const int rows = 8;
            const int cols = 4;

            // Optionally override some defaults on command line
            if (args.Length > 0) 
                customerCount = Int32.Parse(args[0], CultureInfo.CurrentCulture);
            if (args.Length > 1) 
                months = Int32.Parse(args[1], CultureInfo.CurrentCulture);
            if (months < 4) 
                months = 4; // PrintBalance requires at least 4 months

            // Force JIT compilation before timing tests
            const int fewCustomers = 10;
            const int fewMonths = 3;
            AccountRepository smallAccounts = new AccountRepository(fewCustomers, fewMonths, overdraft);
            smallAccounts.AssignRandomTrends(goodBalance, badBalance, variation, random);
            UpdatePredictionsSequential(smallAccounts);
            UpdatePredictionsParallel(smallAccounts);
            UpdatePredictionsPlinq(smallAccounts);

            // Create accounts for timing tests
            AccountRepository accounts = new AccountRepository(customerCount, months, overdraft);
            accounts.AssignRandomTrends(goodBalance, badBalance, variation, random);

            // Print summary of accounts  
            Console.WriteLine();
            Console.WriteLine("{0} customers, {1} months in each account", customerCount, months);

            // Execute sequential and parallel versions, print timings
            Console.WriteLine();
            SampleUtilities.TimedRun(
                () => { UpdatePredictionsSequential(accounts); return customerCount; }, "Sequential");
            SampleUtilities.TimedRun(
                () => { UpdatePredictionsParallel(accounts); return customerCount; }, "  Parallel");
            SampleUtilities.TimedRun(
                () => { UpdatePredictionsPlinq(accounts); return customerCount; }, "     PLINQ");

            // Print a few accounts including predictions and warnings
            accounts.Print(rows, months - cols, cols); // print the last few months

            Console.WriteLine("\nRun complete... press enter to finish.");
            Console.ReadLine();
        }
    }
}

File: parallel-programming-with-ms-dotnet/FSharp/Chapter2/CreditReview/Program.fs
// ===============================================================================
//  Microsoft patterns & practices
//  Parallel Programming Guide
// ===============================================================================
//  Copyright © Microsoft Corporation.  All rights reserved.
//  This code released under the terms of the 
//  Microsoft patterns & practices license (http://parallelpatterns.codeplex.com/license).
// ===============================================================================
module Microsoft.Practices.ParallelGuideSamples.CreditReview.Program

open System
open System.Globalization
open System.Linq
open System.Threading.Tasks
open Microsoft.FSharp.Collections
open Microsoft.Practices.ParallelGuideSamples.Utilities
open Microsoft.Practices.ParallelGuideSamples.CreditReview.Prediction

// ------------------------------------------------------------------------------
// Usage: CreditReview n, optional n is number of customers, use 100,000+ for meaningful timings
// ------------------------------------------------------------------------------

[<EntryPoint>]
let main (args:_[]) = 
    Console.WriteLine("Credit Review Sample\n")
#if DEBUG
    Console.WriteLine("For most accurate timing results, use Release build.\n")
#endif

    // Seed 1 makes runs reproducible
    let random = new Random(1)

    // Defaults for data generation, may override some on command line
    let goodBalance = new Trend(slope = 0.0, intercept = 0.0)
    let badBalance = new Trend(slope = -150.0, intercept = 0.0)
    
    let variation = 100.0
    let overdraft = -1000.0 // Default overdraft limit

    // Printed table of results
    let rows = 8
    let cols = 4

    // Optionally override some defaults on command line
    // For data runs make big enough for significant timing measurements
    let customerCount =
      if args.Length <= 0 then 1000000
      else Int32.Parse(args.[0], CultureInfo.CurrentCulture)

    let months =
      if args.Length <= 1 then 36
      else 
        let m = Int32.Parse(args.[1], CultureInfo.CurrentCulture)
        // PrintBalance requires at least 4 months
        if m < 4 then 4 else m

    // Force JIT compilation before timing tests
    let fewCustomers = 10
    let fewMonths = 3
    let smallAccounts = new AccountRepository(fewCustomers, fewMonths, overdraft)
    smallAccounts.AssignRandomTrends(goodBalance, badBalance, variation, random)

    updatePredictionsSequential(smallAccounts)
    updatePredictionsParallel(smallAccounts)
    updatePredictionsPlinq(smallAccounts)
    updatePredictionsPSeq(smallAccounts)

    // Create accounts for timing tests
    let accounts = new AccountRepository(customerCount, months, overdraft)
    accounts.AssignRandomTrends(goodBalance, badBalance, variation, random)

    // Print summary of accounts  
    Console.WriteLine()
    Console.WriteLine("{0} customers, {1} months in each account", customerCount, months)

    // Execute sequential and parallel versions, print timings
    Console.WriteLine()
    SampleUtilities.TimedRun "Sequential" (fun () ->
        updatePredictionsSequential(accounts)
        customerCount) |> ignore
    SampleUtilities.TimedRun "  Parallel" (fun () ->
        updatePredictionsParallel(accounts)
        customerCount ) |> ignore
    SampleUtilities.TimedRun "     PLINQ" (fun () ->
        updatePredictionsPlinq(accounts) 
        customerCount ) |> ignore
    SampleUtilities.TimedRun "   F# PSeq" (fun () ->
        updatePredictionsPSeq(accounts) 
        customerCount ) |> ignore

    // Print a few accounts including predictions and warnings
    accounts.Print(rows, months - cols, cols) // print the last few months
    Console.WriteLine("\nRun complete... press enter to finish.")
    Console.ReadLine() |> ignore
    0
